amr-wind/wind_energy/ABLStats.cpp

```diff
@@ -0,0 +1,369 @@
+    }
+
+    //Get normal direction and associated stuff
+    const auto& geom = (this->m_sim.repo()).mesh().Geom()[0];
+    amrex::Box const& domain = geom.Domain();
+    const auto dlo = amrex::lbound(domain);
+    const auto dhi = amrex::ubound(domain);
+    switch (m_normal_dir) {
+    case 0:
+        m_ncells_h1 = dhi.y-dlo.y+1;
+        m_ncells_h2 = dhi.z-dlo.z+1;
+        break;
+    case 1:
+        m_ncells_h1 = dhi.x-dlo.x+1;
+        m_ncells_h2 = dhi.z-dlo.z+1;
+        break;
+    case 2:
+        m_ncells_h1 = dhi.x-dlo.x+1;
+        m_ncells_h2 = dhi.y-dlo.y+1;
+        break;
+    }
+    m_dn = geom.CellSize()[m_normal_dir];
+    
+    if (m_out_fmt == "netcdf")
+        prepare_netcdf_file();
+    else
+        prepare_ascii_file();
+        
+}
+
+void ABLStats::post_init_actions()
+{
+    m_pa_vel();
+    m_pa_temp();
+    m_pa_tu();
+    m_pa_uu();
+    m_pa_uuu();
+}
+    
+void ABLStats::post_advance_work()
+{
+    BL_PROFILE("amr-wind::ABLStats::post_advance_work");
+    const auto& time = m_sim.time();
+    const int tidx = time.time_index();
+    // Skip processing if it is not an output timestep
+    if (!(tidx % m_out_freq == 0)) return;
+
+    switch (m_normal_dir) {
+    case 0:
+        compute_zi(YDir(), ZDir());
+        break;
+    case 1:
+        compute_zi(XDir(), ZDir());
+        break;
+    case 2:
+        compute_zi(XDir(), YDir());
+        break;
+    }
+
+    m_pa_vel();
+    m_pa_temp();
+    m_pa_tu();
+    m_pa_uu();
+    m_pa_uuu();
+    
+    process_output();
+}
+
+template <typename h1_dir, typename h2_dir> 
+void ABLStats::compute_zi(const h1_dir& h1Sel, const h2_dir& h2Sel) {
+
+    auto gradT = (this->m_sim.repo()).create_scratch_field(3,m_temperature.num_grow()[0]);
+    compute_gradient(*gradT, m_temperature);
+
+    //Only compute zi using coarsest level
+    
+    amrex::Gpu::ManagedVector<temp_grad> tgrad(m_ncells_h1 * m_ncells_h2);
+    auto * tgrad_ptr = tgrad.data();
+    amrex::Gpu::ManagedVector<temp_grad> tgrad_g(m_ncells_h1 * m_ncells_h2);
+    auto * tgrad_g_ptr = tgrad_g.data();
+    
+    for (amrex::MFIter mfi(m_temperature(0)); mfi.isValid(); ++mfi) {
+      const auto& bx = mfi.tilebox();
+      const auto& gradT_arr = (*gradT)(0).array(mfi);
+      amrex::ParallelFor(
+        bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+        int h1 = h1Sel(i,j,k);
+        int h2 = h2Sel(i,j,k);        
+        if (tgrad[h2*m_ncells_h2 + h1].grad_z
+            < gradT_arr(i,j,k,m_normal_dir)) {
+          tgrad_ptr[h2*m_ncells_h2 + h1].grad_z
+              = gradT_arr(i,j,k,m_normal_dir);
+          tgrad_ptr[h2*m_ncells_h2 + h1].max_grad_loc = (k+0.5)*m_dn;
+        }
+      });
+    }
+
+    MPI_Reduce(tgrad_ptr, tgrad_g_ptr, m_ncells_h1 * m_ncells_h2,
+               MPI_2DOUBLE_PRECISION, MPI_MAXLOC,
+               amrex::ParallelDescriptor::IOProcessorNumber(),
+               amrex::ParallelDescriptor::Communicator() );
```

Comment:
```suggestion
#ifdef AMREX_USE_MPI
    MPI_Reduce(tgrad_ptr, tgrad_g_ptr, m_ncells_h1 * m_ncells_h2,
               MPI_2DOUBLE_PRECISION, MPI_MAXLOC,
               amrex::ParallelDescriptor::IOProcessorNumber(),
               amrex::ParallelDescriptor::Communicator() );
#endif
```

We want to retain the option of not compiling with MPI enabled.